File: video/src/main/java/com/feng/video/db/LocalDataSource.java
package com.feng.video.db;

import android.content.ContentResolver;
import android.content.Context;
import android.database.Cursor;
import android.net.Uri;
import android.os.Environment;
import android.provider.MediaStore;

import com.feng.video.adapter.Item;

import java.io.File;
import java.util.ArrayList;
import java.util.Collections;
import java.util.List;

/**
 * 本地数据源
 */
public class LocalDataSource {

    public static List<Item> getLocalFiles(Context context) {
        final List<File> files = new ArrayList<>();

        //读取存储卡跟目录mp4视频
        File dir = Environment.getExternalStorageDirectory();
        File[] files1 = dir.listFiles((dir1, name) -> name.endsWith("mp4"));

        Collections.addAll(files, files1);
        //读取视频文件
        Uri audioUri = MediaStore.Video.Media.EXTERNAL_CONTENT_URI;
        ContentResolver contentResolver = context.getContentResolver();
        Cursor cursor = contentResolver.query(audioUri, new String[]{MediaStore.Video.VideoColumns.DATA}, null, null, null);
        if (cursor != null && cursor.getCount() != 0) {
            while (cursor.moveToNext()) {
                String path = cursor.getString(cursor.getColumnIndex(MediaStore.Video.Media.DATA));
                int spiltIndex = path.indexOf(".");
                if (spiltIndex != -1) {
                    File file = new File(path);
                    files.add(file);
                }
            }
        }
        if (cursor != null) {
            cursor.close();
        }

        List<Item> items = new ArrayList<>();
        for (File file : files) {
            items.add(new Item(file.getName(), file.getAbsolutePath()));
        }
        return items;
    }
}

File: app/src/main/java/com/feng/player/db/LocalDataSource.kt
package com.feng.player.db

import android.content.Context
import android.os.Environment
import android.provider.MediaStore
import java.io.File

data class Item(val name: String, val path: String)

fun getLocalFiles(context: Context): List<Item> {
    // 使用 HashSet 来去重
    val uniqueFiles = HashSet<File>()

    // 合并文件系统和MediaStore的查询逻辑
    addFilesFromExternalStorage(uniqueFiles)
    addVideosFromMediaStore(context, uniqueFiles)

    // 处理所有找到的文件
    return uniqueFiles.map { Item(it.name, it.absolutePath) }
}

// 从文件系统中添加.mp4文件
private fun addFilesFromExternalStorage(uniqueFiles: HashSet<File>) {
    val externalStorageDirectory = Environment.getExternalStorageDirectory()
    externalStorageDirectory.listFiles { _, name -> name.endsWith(".mp4") }?.forEach { file ->
        uniqueFiles.add(file)
    }
}

// 从MediaStore中添加视频文件
private fun addVideosFromMediaStore(context: Context, uniqueFiles: HashSet<File>) {
    val audioUri = MediaStore.Video.Media.EXTERNAL_CONTENT_URI
    val contentResolver = context.contentResolver
    val cursor = contentResolver.query(audioUri, arrayOf(MediaStore.Video.VideoColumns.DATA), null, null, null)
    cursor?.use {
        while (it.moveToNext()) {
            val path = it.getString(it.getColumnIndex(MediaStore.Video.Media.DATA))
            // 过滤掉非.mp4文件并去重
            if (path.endsWith(".mp4")) {
                val file = File(path)
                uniqueFiles.add(file)
            }
        }
    }
}